pkg/executor/aggregate/agg_spill.go

```diff
@@ -15,15 +15,282 @@
 package aggregate
 
 import (
+	"sync"
 	"sync/atomic"
 
+	"github.com/pingcap/tidb/pkg/executor/aggfuncs"
+	"github.com/pingcap/tidb/pkg/sessionctx"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/disk"
+	"github.com/pingcap/tidb/pkg/util/hack"
 	"github.com/pingcap/tidb/pkg/util/logutil"
 	"github.com/pingcap/tidb/pkg/util/memory"
 	"go.uber.org/zap"
 )
 
-// maxSpillTimes indicates how many times the data can spill at most.
-const maxSpillTimes = 10
+const (
+	// maxSpillTimes indicates how many times the data can spill at most.
+	maxSpillTimes       = 10
+	spillFlag           = 1
+	hasErrorFlag        = 2
+	partialStageFlag    = 3
+	spilledPartitionNum = 256
+	spillTasksDoneFlag  = -1
+
+	notSpillMode = 0
+	spillMode    = 1
+
+	spillLogInfo string = "memory exceeds quota, set aggregate mode to spill-mode"
+)
+
+type parallelHashAggSpillHelper struct {
+	lock struct {
+		mu                         *sync.Mutex
+		cond                       *sync.Cond
+		nextPartitionIdx           int
+		spilledChunksIO            [][]*chunk.DataInDiskByChunks
+		spillTriggered             int32
+		isSpilling                 bool
+		isAllPartialWorkerFinished bool
+	}
+
+	memTracker     *memory.Tracker
+	diskTracker    *disk.Tracker
+	isPartialStage int32
+	hasError       int32
+
+	// These agg functions are partial agg functions that are same with partial workers'.
+	// They only be used for restoring data that are spilled to disk in partial stage.
+	aggFuncsForRestoring []aggfuncs.AggFunc
+
+	// When spill is triggered, all partial workers should stop
+	// and we can ensure this when the spill action gets write lock.
+	syncLock sync.RWMutex
```

Comment:
> Why not move this syncLock into `lock`?

Because `syncLock` is not a lock for avoiding data race, it's just for ensuring that only spill action could run when spill is triggered.